test/blackbox/run_blackbox_tests/test_cases_autoconnect.c

```diff
@@ -0,0 +1,153 @@
+#include <setjmp.h>
+#include <cmocka.h>
+#include <assert.h>
+#include <string.h>
+
+static void test_case_autoconnect(void **state);
+static bool test_steps_mesh_autoconnect(void);
+
+/* State structure for meshlink_blacklist Test Case #1 */
+static black_box_state_t test_mesh_autoconnect_state = {
+	.test_case_name = "test_case_mesh_autoconnect",
+};
+struct sync_flag test_autoconnect_blacklisted = {.mutex  = PTHREAD_MUTEX_INITIALIZER, .cond = PTHREAD_COND_INITIALIZER, .flag = false};
+
+/* Execute meshlink_blacklist Test Case # 1*/
+void test_case_autoconnect(void **state) {
+	execute_test(test_steps_mesh_autoconnect, state);
+}
+
+void callback_logger(meshlink_handle_t *mesh, meshlink_log_level_t level,
+                     const char *text) {
+	(void)mesh;
+	(void)level;
+
+	fprintf((FILE *)mesh->priv, "%s: {%s}\n", mesh->name, text);
+
+	if((check_sync_flag(&test_autoconnect_blacklisted) == true) && (strcmp("m1n2", mesh->name) == 0)) {
+		assert(strcmp(text, "Autoconnect trying to connect to m1n1") != 0);
+	}
+
+}
+
+static void receive(meshlink_handle_t *mesh, meshlink_node_t *src, const void *data, size_t len) {
+	const char *msg = data;
+	assert(len);
+}
+
+static bool bar_reachable;
+
+static void status_cb(meshlink_handle_t *mesh, meshlink_node_t *node, bool reachable) {
+	if(!strcmp(node->name, "bar") && reachable) {
+		bar_reachable = true;
+	}
+}
+
+
+/* Test Steps for meshlink_blacklist Test Case # 1
+
+    Test Steps:
+    1. Open both the node instances
+    2. Join bar node with foo and Send & Receive data
+    3. Blacklist bar and Send & Receive data
+
+    Expected Result:
+    When default blacklist is disabled, foo node should receive data from bar
+    but when enabled foo node should not receive data
+*/
+bool test_steps_mesh_autoconnect(void) {
+	FILE *node1, *node2;
+	char *invite = NULL;
+	meshlink_node_t *node = NULL;
+
+	meshlink_destroy("m1n1");
+	meshlink_destroy("m1n2");
+
+	// Open two new meshlink instance.
+	meshlink_handle_t *mesh1 = meshlink_open("m1n1", "m1n1", "autoconnect", DEV_CLASS_BACKBONE);
+	assert(mesh1 != NULL);
+	assert((node1 = fopen("m1n1.log", "a+")));
+	mesh1->priv = node1;
+	meshlink_set_log_cb(mesh1, TEST_MESHLINK_LOG_LEVEL, callback_logger);
+
+	meshlink_handle_t *mesh2 = meshlink_open("m1n2", "m1n2", "autoconnect", DEV_CLASS_STATIONARY);
+	assert(mesh2 != NULL);
+	assert((node2 = fopen("m1n2.log", "a+")));
+	mesh2->priv = node2;
+	meshlink_set_log_cb(mesh2, TEST_MESHLINK_LOG_LEVEL, callback_logger);
+	meshlink_set_receive_cb(mesh1, receive);
+
+	// Start both instances
+	bar_reachable = false;
+	meshlink_set_node_status_cb(mesh1, status_cb);
+	assert(meshlink_start(mesh1));
+
+	invite = meshlink_invite(mesh1, NULL, "m1n2");
+	assert(invite);
+
+	assert(meshlink_join(mesh2, (const char *)invite));
+
+	assert(meshlink_start(mesh2));
+
+	sleep(2);
+
+	node = meshlink_get_node(mesh2, "m1n1");
+	meshlink_blacklist(mesh2, node);
+	set_sync_flag(&test_autoconnect_blacklisted, true);
+	sleep(30);
```

Comment:
Immediately after the call to `meshlink_join()`, `mesh2` knows about `m1n1`. So you can already blacklist it before calling `meshlink_start(mesh2)`. That way, you don't have to use the `test_autoconnect_blacklisted` flag to ignore the first 2 seconds where it might potentially connect.

But if you want the test to first connect to `m1n1`, then blacklist it, and only then check that it won't autoconnect to it anymore, you should make it explicitly check that it first connected to it before blacklisting, instead of using `sleep(2)`.

Also case, 30 seconds is longer than necessary I think. But I think we can avoid it altogether: if there is no candidate node to connect to, the autoconnect algorithm will log `"* could not find node for initial connect"`. So in this test, you could have the log callback check for that message as well, and if it is received before any message that indicates it does try to autoconnect to `m1n1`, you can exit the test early.